### PR TITLE
tikv: fix `notLeader` in region cache 

### PR DESCRIPTION
**Pull request by lysu**
### What problem does this PR solve? 

fixes bug introduced by  #10256

```
not leader: region_id:8 leader:<id:9 store_id:1 > , ctx: region ID: 8, meta: id:8 end_key:\"t\\200\\000\\000\\000\\000\\000\\000\\005\" region_epoch:<conf_ver:5 version:2 > peers:<id:9 store_id:1 > peers:<id:10 store_id:5 > peers:<id:11 store_id:4 > , peer: id:10 store_id:5 , ad
dr: tidb-cluster-tikv-1.tidb-cluster-tikv-peer.dm-refactor-full-exp328-cat0-tidb-cluster.svc:20160 at 2019-05-22T06:42:46.764390748Z\nnot leader: region_id:8 leader:<id:9 store_id:1 > , ctx: region ID
: 8, meta: id:8 end_key:\"t\\200\\000\\000\\000\\000\\000\\000\\005\" region_epoch:<conf_ver:5 version:2 > peers:<id:9 store_id:1 > peers:<id:10 store_id:5 > peers:<id:11 store_id:4 > , peer: id:10 st
ore_id:5 , addr: tidb-cluster-tikv-1.tidb-cluster-tikv-peer.dm-refactor-full-exp328-cat0-tidb-cluster.svc:20160 at 2019-05-22T06:42:46.775362301Z\nnot leader: region_id:8 leader:<id:9 store_id:1 > , c
tx: region ID: 8, meta: id:8 end_key:\"t\\200\\000\\000\\000\\000\\000\\000\\005\" region_epoch:<conf_ver:5 version:2 > peers:<id:9 store_id:1 > peers:<id:10 store_id:5 > peers:<id:11 store_id:4 > , p
eer: id:10 store_id:5 , addr: tidb-cluster-tikv-1.tidb-cluster-tikv-peer.dm-refactor-full-exp328-cat0-tidb-cluster.svc:20160 at 2019-05-22T06:42:46.786329643Z"
```

the question is NotLeader response told A1 is leader, but  A1 maybe meet other error(e.g. network error) and be marked as `failed`, when next request arrival, TiDB will try round-robin try other nodes, so request will send to A2, and A2 told NotLeader(A1 is leader) again.

### What is changed and how it works?

when recieve NotLeader, not only switch currentPeer to A1 but also clear up A1's failed marker.

then next request will try A1, until Ax return a NotLeader with NULL is leader.

### Check List 

Tests 

 - Unit test
 - Integration test
 - Manual test (add detailed scripts or steps below)

manual running tpcc/sysbench with write leader change test..

schrodinger boxdetail?boxid=15585290796895

Code changes

 - impl change

Side effects

 - N/A

Related changes

 - N/A

---
This change is [Reviewable](https://reviewable.io/reviews/pingcap/tidb/10572)

### CODE REVIEW

File: store/tikv/region_cache.go [hunk 1]
```diff
@@ -1090,6 +1104,9 @@ retry:
 	if !s.compareAndSwapState(oldState, state) {
 		goto retry
 	}
+	if notifyCheckCh == nil {
```
Comment (tiancaiamao) [inline store/tikv/region_cache.go]: When will `notifyCheckCh` be nil?

File: store/tikv/region_cache.go [hunk 2]
```diff
@@ -849,20 +856,27 @@ func (r *Region) EndKey() []byte {
 
 // switchWorkStore switches current store to the one on specific store. It returns
 // false if no peer matches the storeID.
-func (c *RegionCache) switchWorkStore(r *Region, targetStoreID uint64) (switchToTarget bool) {
+func (c *RegionCache) switchWorkStore(r *Region, targetStoreID uint64) {
+	leaderIdx, _ := c.getPeerStoreIndex(r, targetStoreID)
```
Comment (tiancaiamao) [inline store/tikv/region_cache.go]: found is not used?

File: store/tikv/region_cache.go [hunk 3]
```diff
@@ -439,19 +442,27 @@ func (c *RegionCache) InvalidateCachedRegion(id RegionVerID) {
...
+			zap.Uint64("leaderStoreID", leader.GetStoreId()))
+		return
+	}
+
+	if leader == nil {
+		c.switchNext(r, currentPeerIdx)
 		return
 	}
+
+	leaderStoreID := leader.GetStoreId()
```
Comment (coocood) [inline store/tikv/region_cache.go]: If leaderStoreID is 0, we don't need to switchWorkStore.

File: store/tikv/region_cache.go [hunk 4]
```diff
@@ -849,20 +783,49 @@ func (r *Region) EndKey() []byte {
...
+
+func (c *RegionCache) switchNext(r *Region, currentPeerIdx int) {
+	regionStore := r.getStore()
+	if int(regionStore.workStoreIdx) != currentPeerIdx {
+		return
+	}
+	nextIdx := (currentPeerIdx + 1) % len(regionStore.stores)
+	newRegionStore := regionStore.clone()
+	newRegionStore.workStoreIdx = int32(nextIdx)
+	newRegionStore.attemptAfterLoad++
```
Comment (coocood) [inline store/tikv/region_cache.go]: The `attemptAfterLoad` can be removed.

File: store/tikv/region_cache.go (hunk 2 above)
Comment (lysu) [reply store/tikv/region_cache.go]: this has be change in new commit

File: store/tikv/region_cache.go [hunk 5]
```diff
@@ -724,14 +659,10 @@ func (c *RegionCache) getStoreByStoreID(storeID uint64) (store *Store) {
...
 // OnSendRequestFail is used for clearing cache when a tikv server does not respond.
 func (c *RegionCache) OnSendRequestFail(ctx *RPCContext, err error) {
-	failedStoreID := ctx.Store.storeID
-	c.storeMu.RLock()
-	store, exists := c.storeMu.stores[failedStoreID]
-	c.storeMu.RUnlock()
-	if !exists {
-		return
+	r := c.getCachedRegionWithRLock(ctx.Region)
+	if r == nil {
```
Comment (coocood) [inline store/tikv/region_cache.go]: if r != nil

File: store/tikv/region_request.go [hunk 6]
```diff
@@ -177,7 +172,18 @@ func (s *RegionRequestSender) onSendFail(bo *Backoffer, ctx *RPCContext, err err
...
+	r := s.regionCache.getCachedRegionWithRLock(ctx.Region)
+	if r != nil {
+		s.regionCache.switchNextPeer(r, ctx.PeerIdx)
+
+		if s.failStoreIDs == nil {
+			s.failStoreIDs = make(map[uint64]struct{})
+		}
+		s.failStoreIDs[ctx.Store.storeID] = struct{}{}
+		if len(s.failStoreIDs) == len(ctx.Meta.Peers) {
+			r.scheduleReload()
```
Comment (coocood) [inline store/tikv/region_request.go]: We need to clear the `failStoreIDs` after scheduleReload.

File: store/tikv/region_request.go [hunk 7]
```diff
@@ -177,7 +172,18 @@ func (s *RegionRequestSender) onSendFail(bo *Backoffer, ctx *RPCContext, err err
 		}
 	}
 
-	s.regionCache.OnSendRequestFail(ctx, err)
+	r := s.regionCache.getCachedRegionWithRLock(ctx.Region)
+	if r != nil {
+		s.regionCache.switchNextPeer(r, ctx.PeerIdx)
+
```
Comment (coocood) [inline store/tikv/region_request.go]: We need to add comments about this logic.

**Comment by coocood** [comment]
`reloadRegionThreshold` can be removed

File: store/tikv/region_cache.go [hunk 8]
```diff
@@ -439,19 +440,26 @@ func (c *RegionCache) InvalidateCachedRegion(id RegionVerID) {
...
+		c.switchNextPeer(r, currentPeerIdx)
+		return
+	}
+
+	if !c.switchToPeer(r, leaderStoreID) {
 		logutil.Logger(context.Background()).Debug("regionCache: cannot find peer when updating leader",
 			zap.Uint64("regionID", regionID.GetID()),
 			zap.Uint64("leaderStoreID", leaderStoreID))
 		r.invalidate()
+		return
```
Comment (jackysp) [inline store/tikv/region_cache.go]: no need to return

Comment (lysu) [reply store/tikv/region_cache.go]: maybe someone will add code after this in future :rofl:

**Comment by jackysp** [comment]
PTAL @coocood

**Comment by coocood** [comment]
LGTM

**Pull request merged**